File: 剑指offer/easy/20210913/spiralOrder.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
            vector<int>res;
            int n=matrix.size();
            if(n==0){
                return res;
            }
            int m=matrix[0].size();
            int left=0,right=m-1,top=0,bottom=n-1;
            while(left<=right&&top<=bottom){
                for(int i=left;i<=right;i++){
                    res.push_back(matrix[top][i]);
                }
                for(int i=top+1;i<=bottom;i++){
                    res.push_back(matrix[i][right]);
                }
                if(left<right&&top<bottom){
                    for(int i=right-1;i>=left;i--){
                        res.push_back(matrix[bottom][i]);
                    }
                    for(int i=bottom-1;i>top;i--){
                        res.push_back(matrix[i][left]);
                    }
                }
                left++;
                right--;
                top++;
                bottom--;
            }
            return res;
    }
};

File: Leetcode/Review/20210924/spiralOrder.go
package _0210924

func spiralOrder(matrix [][]int) []int {
	n := len(matrix)
	if n == 0 {
		return []int{}
	}
	m := len(matrix[0])
	res := []int{}
	left, right, up, down := 0, m-1, 0, n-1
	for left <= right && up <= down {
		for i := left; i <= right; i++ {
			res = append(res, matrix[up][i])
		}

		for i := up + 1; i <= down; i++ {
			res = append(res, matrix[i][right])
		}

		if left < right && up < down {
			for i := right - 1; i >= left; i-- {
				res = append(res, matrix[down][i])
			}
			for i := down - 1; i > up; i-- {
				res = append(res, matrix[i][left])
			}
		}
		left, right, up, down = left+1, right-1, up+1, down-1
	}
	return res
}
